Task/59 Validate references

ghillairet opened the pull request:
This PR adds support for JSON reference validation inside YAML documents.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Tentative to support file external to workspace by using EFS.

ghillairet committed:
[#59] Validate references

This commit adds the capability to the validator to check for wrong references ($ref keyword).

ghillairet committed:
Merge branch 'task/120' into task/59

ghillairet committed:
[#59] Validate references

ghillairet committed:
[#59] Validate references

This commit includes a preliminary API for dealing with JSON references, notably to validate and resolve JSON references
inside a JSON/YAML document. The API can be found in package com.reprezen.swagedit.json.references.

ghillairet committed:
[#59] Validate references

ghillairet committed:
Merge branch 'master' into task/59

# Conflicts:
#	com.reprezen.swagedit/src/com/reprezen/swagedit/editor/SwaggerEditor.java
#	com.reprezen.swagedit/src/com/reprezen/swagedit/messages.properties

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:
```diff
@@ -49,8 +50,15 @@
+	}
+	
+	public Validator() {
+		this.referenceValidator = new JsonReferenceValidator(new JsonReferenceFactory());
```
Why don't we call `this(new JsonReferenceValidator(new JsonReferenceFactory()))` to make it DRYer?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/DocumentUtils.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	public static SwaggerDocument getDocument(IPath path) {
+		if (path == null || !path.getFileExtension().matches("ya?ml")) {
```
The name is confusing, it's really `getSwaggerDocument` because it return null for any non-Swagger  (and even non-YAML) files

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/DocumentUtils.java:
```diff
@@ -0,0 +1,109 @@
+				.getRoot();
+
+		URI relativize = root.getLocationURI().relativize(uri);
+		IPath path = new Path(relativize.getPath());
```
Does it make sense to call `getWorkspaceFile(IPath path)` here?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/DocumentUtils.java:
```diff
@@ -0,0 +1,109 @@
+		try {
+			return root.getFile(path);
+		} catch (Exception e) {
+			return null;
```
Also, I would suggest using `Optional` here, but it's up to you

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/hyperlinks/JsonReferenceHyperlinkDetector.java:
```diff
@@ -42,109 +35,56 @@
-
-		return doc;
+	protected FileEditorInput getActiveEditor() {
+		return DocumentUtils.getActiveEditorInput();
```
Do we need this method if it simply returns `DocumentUtils.getActiveEditorInput()`?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/JsonUtil.java:
```diff
@@ -10,37 +10,30 @@
+		}
 
-		JsonNode found = null;
+		// TODO Make use of JSONReference
```
Do we still want to make use of JSONReference? If yes, let's create an issue for this. If no, we can remove the TODO

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/JsonUtil.java:
```diff
@@ -49,21 +42,52 @@ public static SchemaDefinition getReference(JsonNode document, JsonNode refNode)
+	}
 
-		return new SchemaDefinition(document, found != null ? found : refNode, lastKey);
+	public static JsonPointer getPointer(JsonNode ref) {
```
`getPointer` and `asPointer` share the following code:

``` java
        if (asText.startsWith("#")) {
            asText = asText.substring(1);
        }

        return JsonPointer.compile(asText);
```

It can be reused

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/JsonUtil.java:
```diff
@@ -49,21 +42,52 @@ public static SchemaDefinition getReference(JsonNode document, JsonNode refNode)
+		return new SchemaDefinition(document, !found.isMissingNode() ? found : refNode, description);
+	}
+
+	public static boolean isPointer(String ptr) {
```
`isPointer` is never used, but that's OK

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/references/JsonReference.java:
```diff
@@ -0,0 +1,153 @@
+ * 
+ * A JSON reference is made of a valid URI and of a JSON Pointer (https://tools.ietf.org/html/rfc6901). 
+ *
+ */
```
Thanks!

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/references/JsonReference.java:
```diff
@@ -0,0 +1,153 @@
+ */
+public class JsonReference {
+
+	public static final String PROPERTY = "$ref";
```
Nice! Can we use this constant in other  places?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/json/references/JsonReference.java:
```diff
@@ -0,0 +1,153 @@
+	private final JsonPointer pointer;
+	private final boolean absolute;
+	private final boolean local;
+	private final Object source;
```
It's so good to see final fields.

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/validation/ReferenceValidatorTest.xtend:
```diff
@@ -0,0 +1,274 @@
+
+		document.set(content)
+		val baseURI = new URI(null, null, null)
+		val otherURI = URI.create("other.yaml")
```
Nice

tfesenko committed:
#59 - fix NPE

tfesenko committed:
ZEN-2716 RAPID-ML Editor Freezes caused by
java.lang.IllegalArgumentException: Illegal character in path at index
9: pathItems\ropeSwagger_address_fulldetails_smartystreets.yaml
	at java.net.URI.create(Unknown Source)
	at java.net.URI.resolve(Unknown Source)
	at
com.reprezen.swagedit.editor.hyperlinks.JsonReferenceHyperlinkDetector.getFile(JsonReferenceHyperlinkDetector.java:114)
	at
com.reprezen.swagedit.editor.hyperlinks.JsonReferenceHyperlinkDetector.doDetectExternalLink(JsonReferenceHyperlinkDetector.java:71)
	at
com.reprezen.swagedit.editor.hyperlinks.JsonReferenceHyperlinkDetector.doDetect(JsonReferenceHyperlinkDetector.java:55)

tfesenko committed:
#59 - fix navigation to externally defined elements (not the same file
as the reference)
Original code works well for cases when referenced files were in the
default workspace location. It did not work for other cases, e.g. when a
project is imported. This fixes it and also simplifies the code by using
standard library.

tfesenko committed:
#59 - fix phrasing for invalid references

tedepstein commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/hyperlinks/JsonReferenceHyperlinkDetector.java:
```diff
@@ -58,13 +58,19 @@ protected boolean canDetect(String basePath) {
-			}
-		}
+        } else {
+            URI resolved;
```
Why are we seeing a difference in the indent level?

This was a red herring w.r.t. to the referenced bug, ZEN-2716.  But it's clearly improved, more robust code that will avoid the unrelated issue observed in the error log.

tedepstein commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/DocumentUtils.java:
```diff
@@ -74,14 +74,7 @@ public static IFile getWorkspaceFile(URI uri) {
-			return null;
-		}
+		return root.getFileForLocation(new Path(uri.getPath()));
 	}
```
Any idea why the previous code was catching `exception`?  Is there a need to do something similar here, or does getFileForLocation manage the expected exception (whatever exception might have been anticipated in the previous code)?

tedepstein replied:
Documentation for the [`IWorkspaceRoot#getFileForLocation`](http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcore%2Fresources%2FIWorkspaceRoot.html) says, " The path should be absolute; a relative path will be treated as absolute."

The previous code seems to anticipate an absolute path, thus the call to `relativize`.  But the new implementation appears to _require_ an absolute URI.  We should document that precondition in the Javadoc and maybe check to make sure it is met; or else coerce to an absolute path.

tfesenko committed:
#59 - simplify getWorkspaceFile(), add javadocs
